In [3]:
] add "QuadGK"

    Updating registry at `~/.julia/registries/General`
   Resolving package versions...
Updating `~/.julia/environments/v1.6/Project.toml`
  [1fd47b50] + QuadGK v2.4.1
No Changes to `~/.julia/environments/v1.6/Manifest.toml`
Precompiling project... (tip: to disable auto-precompilation set ENV["JULIA_PKG_PRECOMPILE_AUTO"]=0)
  ✓ HTTP
  ✓ GR
  ✓ Plots
  ✓ StatsPlots
4 dependencies successfully precompiled (167 already precompiled)


In [6]:
] add "Calculus"

   Resolving package versions...
   Installed Calculus ─ v0.5.1
Updating `~/.julia/environments/v1.6/Project.toml`
  [49dc2e85] + Calculus v0.5.1
Updating `~/.julia/environments/v1.6/Manifest.toml`
  [49dc2e85] + Calculus v0.5.1
Precompiling project... (tip: to disable auto-precompilation set ENV["JULIA_PKG_PRECOMPILE_AUTO"]=0)
  ✓ Calculus
1 dependency successfully precompiled (171 already precompiled)


In [8]:
] add "ForwardDiff"

   Resolving package versions...
Updating `~/.julia/environments/v1.6/Project.toml`
  [f6369f11] + ForwardDiff v0.10.12
No Changes to `~/.julia/environments/v1.6/Manifest.toml`


In [1]:
include("./aux.jl")
include("./a2x.jl")
include("./a3x.jl");

# Finite Differences. Black-Scholes PDE.

## $\S 7.2$ Exercises

**Problem 10**: Asked to consider the following second order ODE:

$$
    3x^2y''(x) - xy'(x) + y(x) = 0, \forall x \in [0, 1] \tag{7.2.1} \\
    y(0) = 1;\qquad y(1) = \frac{1}{2}\qquad\qquad\qquad
$$

I.$\qquad$ Asked to partition the interval [0, 1] into $n$ equal intervals, corresponding to nodes $x_i = ih, i = 0 : n$, where $h = \frac{1}{n}$. Asked to write the finite difference discretization of the ODE at each node, $x_i, i = 1 : (n - 1)$, using central finite difference approximations for both $y'(x)$ and $y''(x)$.

II.$\qquad$ If $n = 6$, from the boundary conditions, $y_0 = 1$ and $y_6 = \frac{1}{2}$. The finite difference discretization scheme will have five equations, which can be written as a $5 \times 5$ linear system $AY = b$. Asked to find $A$ and $b$.

**Solution**:

I.$\qquad$ Asked to let $x_i = ih, i = 0 : n$, where $h = \frac{1}{n}$. Searching for $y_0, y_1, \ldots, y_n$ such that $y_i$ is an approximate value of $y(x_i), \forall\ i \in [0, n]$.

Writing the ODE (7.2.1) at each interior node $x_i = ih$, $i = 1 : (n - 1)$, to obtain

$$
    3x^2y''(x_i) - x_iy'(x_i) + y(x_i) = 0, \forall\ i = 1 : (n - 1)\tag{7.2.2}.
$$

Substituting the second order central difference approximations for $y''(x_i)$ and $y'(x_i)$, respectively; i.e.,

$$
    y''(x_i)\quad = \quad \frac{y(x_{i + 1}) - 2y(x_i) + y(x_{i - 1})}{h^2} + O(h^2) \\
    y'(x_i)\quad = \quad \frac{y(x_{i+1}) - y(x_{i-1})}{2h} + O(h^2)\qquad\quad
$$

into (7.2.2), to obtain the following second order finite difference discretization of (7.2.1), using the approximate values $y_i$ for the exact values $y(x_i)$, for $i = 0 : n$, and ignoring the $O(h^2)$ term:

$$
    3i^2h^2\ \frac{y_{i+1} - 2y_i + y_{i-1}}{h^2} - ih\frac{y_{i+1} - y_{i-1}}{2hy} + y_i = 0, \forall\ i = 1 : (n - 1),
$$

From the boundary conditions, we find that $y_0 = 1$ and $y_n = \frac{1}{2}$.

II.$\qquad$ For $n = 6$, the finite difference discretization above of the ODE (7.2.1) can be written in matrix form as

$$
    A\ Y = b,
$$

where $A$ is a tridiagonal $5 \times 5$ matrix given by

$$
    \qquad A(i, i)\qquad = -(6^2 - 1),\ \forall\ i = 1 : 5; \\
    A(i, i-1)\qquad = 3i^2 + \frac{i}{2},\ \forall\ i = 2 : 5; \\
    A(i, i+1)\qquad = 3i^2 - \frac{1}{2},\ \forall\ i = 1 : 4,
$$

i.e.,

In [2]:
A = reshape(zeros(25), 5, 5)
[A[i, i] = -(6i^2 - 1) for i in 1:5]; [A[i, i - 1] = 3i^2 + i / 2 for i in 2:5]; [A[i, i + 1] = 3i^2 - i / 2 for i in 1:4]; A

5×5 Array{Float64,2}:
 -5.0    2.5    0.0    0.0     0.0
 13.0  -23.0   11.0    0.0     0.0
  0.0   28.5  -53.0   25.5     0.0
  0.0    0.0   50.0  -95.0    46.0
  0.0    0.0    0.0   77.5  -149.0

or

$$
    \left(\begin{matrix}
        -5 & 2.5 & 0 & 0 & 0 \\
        13 & -23 & 11 & 0 & 0 \\
        0 & 28.5 & -53 & 25.5 & 0 \\
        0 & 0 & 50 & -95 & 46 \\
        0 & 0 & 0 & 77.5 & -149
    \end{matrix}\right),
$$

(I found this easier to do in Julia, to be honest.)

and $Y$ and $b$ are the following column vectors:

$$
    Y = \left(\begin{matrix}
                        y_1 \\
                        y_2 \\
                        y_3 \\
                        y_4 \\
                        y_5
        \end{matrix}\right);\qquad b = \left(\begin{matrix}
                                                        -3.5 \\
                                                        0 \\
                                                        0 \\
                                                        0 \\
                                                        -36.25
                                              \end{matrix}\right)
$$

I took the exercise further by solving for $Y$.

In [7]:
b = [-3.5, 0, 0, 0, -36.25]; A \ b

5-element Array{Float64,1}:
 1.3853112213946492
 1.3706224427892995
 1.2286609369112227
 1.0218152955607753
 0.7747697007111416

# Multivariable Calculus. Extremum Points. Barrier Options. Optimality.

## $\S 8.2$ Exercises

**Problem 3**: Asked to compute the integral of the function $f(x, y) = x^2 - 2y$ on the region bounded by the parabola $y = (x + 1)^2 and the line $y = 5x - 1$.

**Solution**: The integration domain $D$ is defined within ${(x + 1)^2, 5x - 1}$. The points where the parabola $(x + 1)^2$ and the line $5x - 1$ meet are at $x = 1$ and $x = 2$, while $(x + 1)^2 \leq 5x - 1$ at $x \in (1, 2)$ as per the following:

$(x + 1)^2 = 5x - 1 => x^2 + 2x + 1 = 5x - 1 => x^2 - 3x + 2 = 0 => (x - 1)(x - 2) = 0$

Then we solve for the $0$-intercepts, and get $1$ and $2$. Something similar is done for the $\leq$ case. I solved this using the double integral:

$$
    \int\int_D f(x, y)dxdy = \int_1^2 \left(\int_{(x + 1)^2}^{5x - 1} (x^2 - 2y)dy\right) dx
$$

I realized I needed to add `QuadGK`, since I did not have it in the package collection somehow.

In [2]:
] add QuadGK

   Updating registry at `~/.julia/registries/General`


   Updating git-repo `https://github.com/JuliaRegistries/General.git`


Fetching: [========================================>]  100.0 %

  Resolving package versions...
  Installed Compat ───────────── v3.14.0
  Installed QuadGK ───────────── v2.4.1
  Installed OrderedCollections ─ v1.3.0
  Installed DataStructures ───── v0.18.2
   Updating `~/.julia/environments/v1.5/Project.toml`
   1fd47b50 + QuadGK v2.4.1
   Updating `~/.julia/environments/v1.5/Manifest.toml`
   34da2185 + Compat v3.14.0
   864edb3b + DataStructures v0.18.2
   bac558e1 + OrderedCollections v1.3.0
   1fd47b50 + QuadGK v2.4.1
   8bb1440f + DelimitedFiles
   37e2e46d + LinearAlgebra
   1a1011a3 + SharedArrays
   2f01184e + SparseArrays
   10745b16 + Statistics


In [4]:
?quadgk

search: quadgk quadgk! QuadGK



```
quadgk(f, a,b,c...; rtol=sqrt(eps), atol=0, maxevals=10^7, order=7, norm=norm)
```

Numerically integrate the function `f(x)` from `a` to `b`, and optionally over additional intervals `b` to `c` and so on. Keyword options include a relative error tolerance `rtol` (if `atol==0`, defaults to `sqrt(eps)` in the precision of the endpoints), an absolute error tolerance `atol` (defaults to 0), a maximum number of function evaluations `maxevals` (defaults to `10^7`), and the `order` of the integration rule (defaults to 7).

Returns a pair `(I,E)` of the estimated integral `I` and an estimated upper bound on the absolute error `E`. If `maxevals` is not exceeded then `E <= max(atol, rtol*norm(I))` will hold. (Note that it is useful to specify a positive `atol` in cases where `norm(I)` may be zero.)

The endpoints `a` et cetera can also be complex (in which case the integral is performed over straight-line segments in the complex plane). If the endpoints are `BigFloat`, then the integration will be performed in `BigFloat` precision as well.

!!! note
    It is advisable to increase the integration `order` in rough proportion to the precision, for smooth integrands.


More generally, the precision is set by the precision of the integration endpoints (promoted to floating-point types).

The integrand `f(x)` can return any numeric scalar, vector, or matrix type, or in fact any type supporting `+`, `-`, multiplication by real values, and a `norm` (i.e., any normed vector space). Alternatively, a different norm can be specified by passing a `norm`-like function as the `norm` keyword argument (which defaults to `norm`).

!!! note
    Only one-dimensional integrals are provided by this function. For multi-dimensional integration (cubature), there are many different algorithms (often much better than simple nested 1d integrals) and the optimal choice tends to be very problem-dependent. See the Julia external-package listing for available algorithms for multidimensional integration or other specialized tasks (such as integrals of highly oscillatory or singular functions).


The algorithm is an adaptive Gauss-Kronrod integration technique: the integral in each interval is estimated using a Kronrod rule (`2*order+1` points) and the error is estimated using an embedded Gauss rule (`order` points). The interval with the largest error is then subdivided into two intervals and the process is repeated until the desired error tolerance is achieved.

These quadrature rules work best for smooth functions within each interval, so if your function has a known discontinuity or other singularity, it is best to subdivide your interval to put the singularity at an endpoint. For example, if `f` has a discontinuity at `x=0.7` and you want to integrate from 0 to 1, you should use `quadgk(f, 0,0.7,1)` to subdivide the interval at the point of discontinuity. The integrand is never evaluated exactly at the endpoints of the intervals, so it is possible to integrate functions that diverge at the endpoints as long as the singularity is integrable (for example, a `log(x)` or `1/sqrt(x)` singularity).

For real-valued endpoints, the starting and/or ending points may be infinite. (A coordinate transformation is performed internally to map the infinite interval to a finite one.)


I could not figure out how to integrate over "variable limits of integration". Maybe `SymPy`, the computer algebra system, may have a solution, since `QuadGK` requires fixed limits of integration, and integrating over a domain is not working in Julia.

In [21]:
] add SymPy

  Resolving package versions...
  Installed CompilerSupportLibraries_jll ─ v0.3.3+0
  Installed OpenSpecFun_jll ────────────── v0.5.3+3
  Installed RecipesBase ────────────────── v1.0.2
  Installed SpecialFunctions ───────────── v0.10.3
  Installed PyCall ─────────────────────── v1.91.4
  Installed MacroTools ─────────────────── v0.5.5
  Installed SymPy ──────────────────────── v1.0.27
######################################################################### 100.0%#################                                  57.3%
######################################################################### 100.0%
   Updating `~/.julia/environments/v1.5/Project.toml`
   24249f21 + SymPy v1.0.27
   Updating `~/.julia/environments/v1.5/Manifest.toml`
   e66e0078 + CompilerSupportLibraries_jll v0.3.3+0
   1914dd2f + MacroTools v0.5.5
   efe28fd5 + OpenSpecFun_jll v0.5.3+3
   438e738f + PyCall v1.91.4
   3cdcf5f2 + RecipesBase v1.0.2
   276daf66 + SpecialFunctions v0.10.3
   24249f21 + SymPy v1.0.27
   B

In case I need `PyCall`, also adding this package.

In [38]:
] add PyCall

  Resolving package versions...
   Updating `~/.julia/environments/v1.5/Project.toml`
   438e738f + PyCall v1.91.4
   Updating `~/.julia/environments/v1.5/Manifest.toml`
   Updating empty diff


In [43]:
using PyCall, SymPy

@vars h
f(x) = x^2 + 2y
ex1 = SymPy.exp((z + 1)^2)
ex2 = SymPy.exp(5z - 1)
ex = integrate(z^2 - 2y, y, ex1, ex2)

ArgumentError: ArgumentError: ref of NULL PyObject

This would not work either. I found an `R` tool called `pracma` which might work, but before using it, I must install `RCall`.

In [48]:
] add RCall

  Resolving package versions...
  Installed SortingAlgorithms ─────────── v0.3.1
  Installed Requires ──────────────────── v1.0.1
  Installed Reexport ──────────────────── v0.2.0
  Installed ShiftedArrays ─────────────── v1.0.0
  Installed StatsModels ───────────────── v0.6.13
  Installed CategoricalArrays ─────────── v0.8.1
  Installed RCall ─────────────────────── v0.13.7
  Installed InvertedIndices ───────────── v1.0.0
  Installed DataAPI ───────────────────── v1.3.0
  Installed Rmath ─────────────────────── v0.6.1
  Installed TableTraits ───────────────── v1.0.0
  Installed Tables ────────────────────── v1.0.5
  Installed StatsFuns ─────────────────── v0.9.5
  Installed DataFrames ────────────────── v0.21.7
  Installed Distributions ─────────────── v0.23.10
  Installed IteratorInterfaceExtensions ─ v1.0.0
  Installed Missings ──────────────────── v0.4.4
  Installed DataValueInterfaces ───────── v1.0.0
  Installed Rmath_jll ─────────────────── v0.2.2+1
  Installed PooledArrays ─────

Using `RCall` and installing `pracma`.

In [51]:
using RCall

R"""
    install.packages('pracma')
"""

┌ Warning: RCall.jl: Installing package into ‘/usr/local/lib/R/site-library’
│ (as ‘lib’ is unspecified)
│ Warning: unable to access index for repository https://cloud.r-project.org/src/contrib:
│   cannot open URL 'https://cloud.r-project.org/src/contrib/PACKAGES'
│ Warning: package ‘pracma’ is not available (for R version 3.6.3)
└ @ RCall /root/.julia/packages/RCall/Qzssx/src/io.jl:160


RObject{NilSxp}
NULL


Something happened. `pracma` seems to be unavailable in the current repository. It is unlikely to be a problem with version, since the package is unavailble in the first place, but this is what I tried doing earlier.

In [58]:
R"""
    library(remotes)
    install_version('pracma', "2.2.9")
"""

LoadError: REvalError: Warning: unable to access index for repository https://cloud.r-project.org/src/contrib:
  cannot open URL 'https://cloud.r-project.org/src/contrib/PACKAGES'
Error in package_find_repo(package, repos) : 
  couldn't find package 'pracma'
Calls: install_version -> download_version_url -> package_find_repo

See? The same **couldn't find package `pracma`** error. Trying to install from `git` with my `user` `pass` credentials.

In [70]:
R"""
    remotes::install_git("https://github.com/salma-rodriguez/pracma.git",
                     credentials=git2r::cred_user_pass("salma-rodriguez", "c@trp1ll@r5"))
"""

LoadError: REvalError: Error: Failed to install 'unknown package' from Git:
  Error in 'git2r_remote_ls': the SSL certificate is invalid

I do not know what happened here. I tried fixing the error. The package returned to be *unknown* and the SSL certificate was found to be invalid. Maybe the credentials need to be declared outside?

In [72]:
R"""
    library(git2r)
    creds <- new("cred_user_pass", username = "salma-rodriguez", password = "c@trp1ll@r5")
    remotes::install_git("https://github.com/salma-rodriguez/pracma.git",
                     credentials=creds)
"""

LoadError: REvalError: Error in getClass(Class, where = topenv(parent.frame())) : 
  “cred_user_pass” is not a defined class
Calls: new -> getClass

See? This did not work either. `cred_user_pass` not a defined class, eh? Let us try calling the `available.packages()` function, and see what happens.

In [76]:
R"""
    ap <- available.packages()
    # install_github("https://github.com/salma-rodriguez/pracma")
"""
@rget ap

┌ Warning: RCall.jl: Warning: unable to access index for repository https://cloud.r-project.org/src/contrib:
│   cannot open URL 'https://cloud.r-project.org/src/contrib/PACKAGES'
└ @ RCall /root/.julia/packages/RCall/Qzssx/src/io.jl:160


0×17 Array{String,2}

There seems to be a problem with `https://cloud.r-project.org/src/contrib`. I will have to fix this later. As a last resort, I am trying to install `pracma` via `R-Forge`.

In [78]:
R"""
    install.packages("pracma", repos="http://R-Forge.R-project.org")
"""

* installing *source* package ‘pracma’ ...
** using staged installation
** R
** data
*** moving datasets to lazyload DB
** demo
** byte-compile and prepare package for lazy loading
** help
*** installing help indices
** building package indices
** testing if installed package can be loaded from temporary location
** testing if installed package can be loaded from final location
** testing if installed package keeps a record of temporary installation path
* DONE (pracma)
┌ Warning: RCall.jl: Installing package into ‘/usr/local/lib/R/site-library’
│ (as ‘lib’ is unspecified)
│ trying URL 'http://R-Forge.R-project.org/src/contrib/pracma_2.3.0.tar.gz'
│ Content type 'application/x-gzip' length 372667 bytes (363 KB)
│ ==================================================
│ downloaded 363 KB
│ 
│ 
│ The downloaded source packages are in
│ 	‘/tmp/RtmpuENBnb/downloaded_packages’
└ @ RCall /root/.julia/packages/RCall/Qzssx/src/io.jl:160


RObject{NilSxp}
NULL


This seems to have worked. Now to run some code.

In [132]:
R"""
    library(pracma)
    f <- function(x, y) x^2 - 2 * y
    xmin <- 1; xmax <- 2
    ymin <- function(x) (x + 1)^2; ymax = function(x) 5 * x - 1
    I <- integral2(f, xmin, xmax, ymin, ymax)
    MyQ <- I$Q
"""
I = @rget(MyQ)

-1.7500000000000016

And this is the double integral within the region $R$ specified in the beginning of the problem.

**Problem 14**: The price of a non-dividend-paying asset is lognormally distributed. Asked to assume that the spot price is $40$, the volatility is $30 \%$, and the interest rates are constant at $5\%$. Asked to fill in the Black-Scholes values of the ITM put options in a table. I choose to create an `R` data frame.

In [4]:
] add DataFrames

   Updating registry at `~/.julia/registries/General`


   Updating git-repo `https://github.com/JuliaRegistries/General.git`


    Fetching: [=======

Fetching: [========================================>]  100.0 %                       ]  39.2 %]  50.6 %===========================>             ]  66.0 %8 %>   ]  90.2 %

  Resolving package versions...
   Updating `~/.julia/environments/v1.5/Project.toml`
   a93c6f00 + DataFrames v0.21.7
   Updating `~/.julia/environments/v1.5/Manifest.toml`
   Updating empty diff


In [30]:
using DataFrames

t = 0; S = 40; K = [45, 45, 48, 48, 51, 51]; T = [6, 3, 6, 3, 6, 3]; σ = 0.30; r = 0.05; q = 0

df = DataFrame(
    Option_Type = ["PUT" for I in 1:6],
    Strike = S,
    Maturity = T,
    Value = [black_scholes_approx(t, S, K[I], T[I] / 12, σ, r, q)[2][2] for I in 1:size(df)[1]],
    KminusS = [K[I] - S for I in 1:6]
)

6×5 DataFrame
│ Row │ Option_Type │ Strike │ Maturity │ Value   │ KminusS │
│     │ String      │ Int64  │ Int64    │ Float64 │ Int64   │
├─────┼─────────────┼────────┼──────────┼─────────┼─────────┤
│ 1   │ PUT         │ 40     │ 6        │ 5.81958 │ 5       │
│ 2   │ PUT         │ 40     │ 3        │ 5.3403  │ 5       │
│ 3   │ PUT         │ 40     │ 6        │ 8.03253 │ 8       │
│ 4   │ PUT         │ 40     │ 3        │ 7.8234  │ 8       │
│ 5   │ PUT         │ 40     │ 6        │ 10.4862 │ 11      │
│ 6   │ PUT         │ 40     │ 3        │ 10.5476 │ 11      │

The purpose of this exercise was to show that **deep-in-the money** European put options are lower than the premium $K - S$ (the put options with higher $K - S$ are deep-in-the money). If the spot price is $0$, the value of the option would be $Ke^{-rT}$, which shows that longer dated puts are worth less than short dated ones (one cannot wait until the spot price goes to $0$, because then you have a discounted strike anyway).

# Lagrange Multipliers. Portfolio Optimization

## Minimum Variance Portfolio

**Note**: the following **minimum variance portfolio** example was finished after the **maximum return portfolio** example, but since I must conserve the order of chronology, I am including this one first.

**Example 1**: Find a minimum variance portfolio with $11.5\%$ expected rate of return, if four assets can be traded to set up the portfolio, given the following data on the rates of return of the assets.

**Solution**: Let

$$
    \mu = \left(\begin{matrix}
              0.09\\ 0.12\\ 0.15\\ 0.06
          \end{matrix}\right)\hspace{1em} \text{and}\hspace{1em} M = \left(\begin{matrix}
                                                                               0.04 & -0.03 & 0.02450 & 0 \\
                                                                               -0.03 & 0.09 & 0.02625 & 0 \\
                                                                               0.0245 & 0.02625 & 0.1225 & 0 \\
                                                                               0 & 0 & 0 & 0.0225
                                                                      \end{matrix}\right)
$$

be the vector of expected values of the rates of return and the covariance matrix of the rates of return of the assets, respectively and as computed below.

In [2]:
μp = 0.115
μ = [0.09, 0.12, 0.15, 0.06]
σ = [0.20, 0.30, 0.35, 0.15];
ρ = [
        +1.00, -0.50, 0.35, 0.00,
        -0.50, +1.00, 0.25, 0.00,
        +0.35, +0.25, 1.00, 0.00,
        +0.00, +0.00, 0.00, 1.00
]; ρ = reshape(ρ, 4, 4);

In [3]:
M = VarCov2(μ, σ, ρ)

4×4 Array{Float64,2}:
  0.04    -0.03     0.0245   0.0
 -0.03     0.09     0.02625  0.0
  0.0245   0.02625  0.1225   0.0
  0.0      0.0      0.0      0.0225

I solve the linear system

$$
    \left(\begin{matrix}
        2M & \textbf{1} & \mu \\ \textbf{1}^t & 0 & 0 \\ \mu^t & 0 & 0
    \end{matrix}\right)\ \left( \begin{matrix} w \\ \lambda_1 \\ \lambda_2\end{matrix} \right)\ = \left(\begin{matrix} 0 \\ 1 \\ \mu_p \end{matrix}\right)
$$

to obtain the weights of the minimum variance portfolio, implemented as `__nablaF`, which gives me the solution

In [8]:
μ0 = μ; λ0 = [1, μp]; x = [μ0; λ0]
xp = __nablaF(x, M) \ [0, 0, 0, 0, 1, μp]

6-element Array{Float64,1}:
  0.5474522427456003
  0.4403771370834482
  0.135042272140279
 -0.12287165196932762
  0.06456891807488878
 -0.9839948956044839

The standard deviation of the portfolio, $σ_p = \sqrt{w_0^tM\ w_0}$, is then

In [9]:
sqrt(xp[1:4]'M * xp[1:4])

0.15586932815603413

or $15.586933\%$

## Maximum Return Portfolio

**Example 2**: Find a maximum return portfolio with $25\%$ standard deviation of the rate of return, if four assets can be traded to set up the portfolio, given by the following data on the rates of return of the assets:

$$
    \begin{matrix}
        \mu_1 = 0.09; & \sigma_1 = 0.20; & \rho_{1,2} = -0.5; \\
        \mu_2 = 0.12; & \sigma_2 = 0.30; & \rho_{2,3} = 0.25; \\
        \mu_3 = 0.15; & \sigma_3 = 0.35; & \rho_{1,3} = 0.35; \\
        \mu_4 = 0.06; & \sigma_4 = 0.15; & \rho_{i,4} = 0,\ \forall\ i = 1 : 3
    \end{matrix}
$$

**Solution**: Let

In [3]:
σp = 0.25
μ = [0.09, 0.12, 0.15, 0.06]
σ = [0.20, 0.30, 0.35, 0.15]
ρ = [
        +1.00, -0.50, 0.35, 0.00,
        -0.50, +1.00, 0.25, 0.00,
        +0.35, +0.25, 1.00, 0.00,
        +0.00, +0.00, 0.00, 1.00
]; ρ = reshape(ρ, 4, 4);

and the variance-covariance matrix, which would be computed as on page $283$ of text, i.e.,

$$
    M = \left(\begin{matrix}
        \sigma_1^2 & \sigma_1\sigma_2\rho_{1,2} & \sigma_1\sigma_3\rho_{1,3} & \cdots & \sigma_1\sigma_n\rho_{1,n} \\
        \sigma_1\sigma_2\rho_{1,2} & \sigma_2^2 & \sigma_2\sigma_3\rho_{2,3} & \cdots & \sigma_2\sigma_n\rho_{2,n} \\
        \sigma_1\sigma_3\rho_{1,3} & \sigma_2\sigma_3\rho_{2,3} & \sigma_3^2 & \cdots & \sigma_3\sigma_n\rho_{3,n} \\
        \vdots & \vdots & \vdots & \ddots & \vdots \\
        \sigma_1\sigma_n\rho_{1,n} & \sigma_2\sigma_n\rho_{2,n} & \sigma_3\sigma_n\rho_{3,n} & \cdots & \sigma_n^2
    \end{matrix}\right)
$$

be

In [4]:
M = VarCov2(μ, σ, ρ)

4×4 Matrix{Float64}:
  0.04    -0.03     0.0245   0.0
 -0.03     0.09     0.02625  0.0
  0.0245   0.02625  0.1225   0.0
  0.0      0.0      0.0      0.0225

$\mu$ is the vector of expected returns of the Assets Under Management (AUM) and $M$ is the covariance matrix of the rates of return of the assets, respectively.

Recall that the following condition must be satisfied, to find the maximum return portfolio using the Lagrange multiplier method: $\textbf{1}^tM^{-1}\textbf{1} \neq \frac{1}{\sigma_p^2}$.

Checking whether or not $\textbf{1}^tM^{-1}\textbf{1} \neq \frac{1}{\sigma_p^2}$ is satisfied, assuring that the Lagrange Multiplier Method can be used for this constrained maximization example.

In [48]:
ones(4)' * M ^ -1 * ones(4) != 1 / σp^2

true

The weights of the maximum return portfolio can be found by solving $G(w, \lambda_1, \lambda_2) = 0$, where the function $G(w, \lambda_1, \lambda_2)$ is the transpose of the gradient of the Langrangian function associated to this constrained optimization problem. We obtain that the function $G: \cal R^6 \rightarrow \cal R^6$ is given by

$$
    G(w, \lambda_1, \lambda_2) = \left(\begin{matrix}
                                 \mu + \lambda_1\textbf{1} + 2\lambda_2M\ w \\
                                 \textbf{1}^tw - 1 \\
                                 w^tM\ w - \sigma_p^2 \end{matrix} \right)
$$

In [125]:
σp = 0.25;
G(x) = [μ + x[5] * ones(size(x[1:end-2])[1]) + 2x[6] * M * x[1:end-2]; 
    ones(size(x[1:end-2])[1])'x[1:end-2] - 1; x[1:end-2]'M * x[1:end-2] - 0.625]

G (generic function with 1 method)

In [126]:
x0 \ G(x0)

0.9407916666666666

$\nabla_{(w, \lambda))}\ G(w, \lambda_1, \lambda_2)$ is a function, giving the $6 \times 6$ matrix

$$
    \nabla_{(w, \lambda)}\ G(w, \lambda_1, \lambda_2) = \left(\begin{matrix}
                                                            2\lambda_2 M & \textbf{1} & 2M\ w \\
                                                            \textbf{1}^t & 0 & 0 \\
                                                            2(M\ w)^t & 0 & 0
                                                        \end{matrix}\right)
$$

I tried taking the gradient of $G$ using the standard linear algebra method, but this obviously would not work, due to having nonlinear components. I learned that I must always use **Lagrange Multiplier** method for nonlinear problems. `nDimNewton` needs to be modified for this exercise.

In [121]:
nDimNewton(w0, G)

LoadError: MethodError: no method matching size(::typeof(G))
Closest candidates are:
  size(!Matched::BitArray{1}) at bitarray.jl:99
  size(!Matched::BitArray{1}, !Matched::Integer) at bitarray.jl:102
  size(!Matched::Core.Compiler.StmtRange) at show.jl:1846
  ...

To implement the gradient function above, I wrote a helper function called `__nablaG`.

In [122]:
__nablaG(x0, M)

6×6 Array{Float64,2}:
  0.08    -0.06     0.049    0.0     1.0  0.0345
 -0.06     0.18     0.0525   0.0     1.0  0.08625
  0.049    0.0525   0.245    0.0     1.0  0.17325
  0.0      0.0      0.0      0.045   1.0  0.0225
  1.0      1.0      1.0      1.0     0.0  0.0
  0.0345   0.08625  0.17325  0.0225  0.0  0.0

To get the Lagrange Multiplier method to work for `nDimNewton`, I wrote a new function called `nDimNewtonLagrange`.

As of Thursday, September 24, 2020, I detected a bug in my implementation of `G`.

$\sigma_p$ is the standard deviation of the portfolio, not of the individual assets. $\sigma_p$ is, therefore, the **target risk** that I should have written for `G` above.

I am leaving the output of my previous implementation of `G`, to be kindly reminded of this naive implementation error later.

In [127]:
w0 = [0.5, 0.5, 0.5, 0.5]

nDimNewtonLagrange(w0, G, M)

(9, [1.0401698927536367, 1.3492142023807232, 1.2530895030648683, -2.6424735981992287], [-0.07953920252137467, -0.16431744134393278])

This is what I should have gotten, had I implemented the `G` function correctly, as I am doing now.

In [12]:
G(x) = [μ + x[end-1] * ones(size(x[1:end-2])[1]) + 2x[end] * M * x[1:end-2]; 
    ones(size(x[1:end-2])[1])'x[1:end-2] - 1; x[1:end-2]'M * x[1:end-2] - σp^2]

w0 = [0.5, 0.5, 0.5, 0.5]; λ1 = 1; λ2 = 1

n, w0, λ0 = nDimNewtonLagrange(w0, G, M, λ1 = 1, λ2 = 1)

(7, [0.6268544356233006, 0.5868376017165989, 0.31521727069125743, -0.528909308031157], [-0.07317766900846581, -0.5536621962035316])

The expected return of the portfolio is $\mu_p = \mu^tw_0$ or

In [14]:
μp = μ'w0

0.14238544353390814

### **G** "Implementation" Given in Text

This implementation is directly from the code on page $290$. Again, I got $\sigma$ wrong the first time, but I am including for reference.

In [214]:
G(x) = [
    0.09 + x[5] + 0.080 * x[6] * x[1] - 0.0600 * x[6] * x[2] + 0.0450 * x[6] * x[3],
    0.12 + x[5] - 0.060 * x[6] * x[1] + 0.1800 * x[6] * x[2] + 0.0525 * x[6] * x[3],
    0.15 + x[5] + 0.049 * x[6] * x[1] + 0.0525 * x[6] * x[2] + 0.2450 * x[6] * x[3],
    0.06 + x[5] + 0.090 * x[6] * x[4],
    x[1] + x[2] + x[3] + x[4] - 1,
    0.04 * x[1]^2 + 0.09 * x[2]^2 + 0.1255 * x[3]^2 + 0.0225 * x[4]^2 - 0.06 * x[1] * x[2] + 0.049 * x[1] * x[3] + 0.0525 * x[2] * x[3] - 0.625
];

In [217]:
nDimNewtonLagrange(x0, G, M)

(34, [0.3238585926615457, 1.0439844372601506, 1.6379648130738014, -2.0058078429954973, -0.08489959959938345, -0.1379305288223673])

Implementation with the "correct" value of $\sigma$ in $G$.

In [250]:
G(x) = [
    0.09 + x[5] + 0.080 * x[6] * x[1] - 0.0600 * x[6] * x[2] + 0.0450 * x[6] * x[3],
    0.12 + x[5] - 0.060 * x[6] * x[1] + 0.1800 * x[6] * x[2] + 0.0525 * x[6] * x[3],
    0.15 + x[5] + 0.049 * x[6] * x[1] + 0.0525 * x[6] * x[2] + 0.2450 * x[6] * x[3],
    0.06 + x[5] + 0.090 * x[6] * x[4],
    x[1] + x[2] + x[3] + x[4] - 1,
    0.04 * x[1]^2 + 0.09 * x[2]^2 + 0.1255 * x[3]^2 + 0.0225 * x[4]^2 - 0.06 * x[1] * x[2] + 0.049 * x[1] * x[3] + 0.0525 * x[2] * x[3] - 0.0625
];

I got a different result this time. It is not remotely close to the result I was expecting.

In [252]:
w0 = [0.5, 0.5, 0.5, 0.5]
nDimNewtonLagrange(w0, G, M)

(24, [0.4854920217534955, 0.5271124216975402, 0.39365568587610794, -0.4062601293271437], [-0.07783855289915374, -0.48788011363771805])

## $\S 9.2$ Exercises

**Problem 3**: Asked to consider five tradable assets with the following expected values and standard deviations of the rates of return of the assets:

\begin{matrix}
    \mu_1 = 0.08 & \mu_2 = 0.10 & \mu_3 = 0.13 & \mu_4 = 0.15 & \mu_5 = 0.20 \\
    \sigma_1 = 0.14 & \sigma_2 = 0.18 & \sigma_3 = 0.23 & \sigma_4 = 0.25 & \sigma_5 = 0.35
\end{matrix}

The correlation matrix of the rates of return is

$$\Omega = \left(\begin{matrix}
    1 & -0.3 & 0.4 & 0.25 & -0.2 \\
    -0.3 & 1 & -0.1 & -0.2 & 0.15 \\
    0.4 & -0.1 & 1 & 0.35 & 0.25 \\
    0.25 & -0.2 & 0.35 & 1 & -0.15 \\
    -0.2 & 0.15 & 0.25 & -0.15 & 1
\end{matrix}\right)$$

Assume that it is possible to take both long and short positions of arbitrary size in these assets.

(i) Find the asset allocation for a minimal variance portfolio with $15\%$ expected rate of return and the corresponding minimal standard deviation of the rate of return of the portfolio.

(ii) Find the asset allocation for a maximum expected return portfolio with $25\%$ standard deviation of the rate of return and the corresponding maximal expected rate of return of the portfolio.

**Solution**: Let

In [20]:
μp = 0.15
μ = [0.08, 0.10, 0.13, 0.15, 0.20]; σ = [0.14, 0.18, 0.23, 0.25, 0.35];
ρ = [
    1, -0.3, 0.4, 0.25, -0.2,
    -0.3, 1, -0.1, -0.2, 0.15,
    0.4, -0.1, 1, 0.35, 0.25,
    0.25, -0.2, 0.35, 1, -0.15,
    -0.2, 0.15, 0.25, -0.15, 1
]; ρ = reshape(ρ, 5, 5)

5×5 Matrix{Float64}:
  1.0   -0.3    0.4    0.25  -0.2
 -0.3    1.0   -0.1   -0.2    0.15
  0.4   -0.1    1.0    0.35   0.25
  0.25  -0.2    0.35   1.0   -0.15
 -0.2    0.15   0.25  -0.15   1.0

be the vectors of expected rates of return and of the standard deviation of the rates of return of the given assets, respectively. The correlation matrix, $\rho$, is also given.

The covariance matrix $M$ of the rates of return of the assets is given by $M(i, j) = \sigma(i)\sigma(j)\rho_{i, j}, \forall i, j = 1:5$, where $\rho = \omega(i, j)$ is the correlation of the rates of return of the assets $i$ and $j$. Basically, the matrix $M = \text{diag}(\sigma)\ \Omega\ \text{diag}({\sigma})$. Then,

In [35]:
M = VarCov2(μ, σ, ρ)

5×5 Matrix{Float64}:
  0.0196   -0.00756   0.01288   0.00875  -0.0098
 -0.00756   0.0324   -0.00414  -0.009     0.00945
  0.01288  -0.00414   0.0529    0.02012   0.02012
  0.00875  -0.009     0.02012   0.0625   -0.01312
 -0.0098    0.00945   0.02012  -0.01312   0.1225

#### For the minimum variance portfolio...

In [36]:
μ0 = μ; λ0 = [1, μp]; x = [μ0; λ0]
xp = __nablaF(x, M) \ [0, 0, 0, 0, 0, 1, μp]

7-element Vector{Float64}:
  0.0672828417168285
  0.21731272173207297
 -0.002110778934250173
  0.406850826923416
  0.3106643885619327
  0.04488000427746616
 -0.5651027138772067

and remember that $σ_p = \sqrt{w_0^tMw_0}$

In [34]:
σp = √(xp[1:end-2]'M * xp[1:end-2])

0.1412186297981163

#### For the maximum return portfolio...

In [38]:
σp = 0.25; G(x) = [μ + x[end-1] * ones(size(x[1:end-2])[1]) + 2x[end] * M * x[1:end-2]; 
                   ones(size(x[1:end-2])[1])'x[1:end-2] - 1; 
                   x[1:end-2]'M * x[1:end-2] - σp^2]

G (generic function with 1 method)

Is the **Lagrange Multiplier Method (LMM)** condition met?

In [39]:
ones(5)' * M ^ -1 * ones(5) != 1 / σp

true

We may continue.

In [40]:
w0 = [0.5, 0.5, 0.5, 0.5, 0.5];

n, w0, λ0 = nDimNewtonLagrange(w0, G, M, λ1 = 1, λ2 = 1)

(7, [-0.39559845940111615, 0.09609158418614616, 0.04606428037709733, 0.7097907315309685, 0.5436518633069043], [-0.09200673706952392, -0.8571350674962558])

**Final Note**: Although I did not get the results precisely as written in the text, I did find a few errors in my original code. I was using the portfolio weights as if it was $4$, giving me $\lambda_1$ at position $5$ in the $x_0$ array and $\lambda_2$ at position $6$. **Reminder to generalize each and every function!**

The expected return of the portfolio is given by $μ^tw_0$

In [18]:
μ'w0

0.19914862050657411